### PR TITLE
Stride order input enablement

### PR DESCRIPTION
**Pull request by jjsjann123**
1. Adding stride_order in TensorRecord;
2. Adding stride_order in both python API define_tensor;

### CODE REVIEW

File: csrc/python_frontend/python_bindings.cpp [hunk 1]
```diff
@@ -666,13 +669,17 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+	    
             self.defineRecord(new TensorRecord(
                 {self.recordingState(out())},
                 std::move(dim_sizes),
-                computeContiguity(sizes, strides),
+		contiguity,
                 dtype,
-                is_cpu));
+                is_cpu,
+		stride_order));
```
Comment (jjsjann123) [inline csrc/python_frontend/python_bindings.cpp:682]: NoteToMyself: The change here needs to be picked up by integration's cache logic.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
I feel I should have add some python tests to verify this. @kevinstephano this is the point where having some ability to query kernel heuristics would be useful. i.e. I need to assert on vectorization width to verify that allocation domain has worked.

File: csrc/python_frontend/fusion_record.h [hunk 2]
```diff
@@ -1168,14 +1168,16 @@ struct TensorRecord : RecordFunctor {
       std::vector<int64_t> _shape,
       std::vector<std::optional<bool>> _contiguity,
       PrimDataType _dtype,
-      bool _is_cpu = false)
+      bool _is_cpu = false,
+      std::vector<int64_t> _stride_order = {})
```
Comment (jjsjann123) [inline csrc/python_frontend/fusion_record.h:1172]: Note to myself. API change here bumps version number.

File: csrc/python_frontend/fusion_record.h [hunk 3]
```diff
@@ -1264,6 +1278,17 @@ struct TensorRecord : RecordFunctor {
       NVF_CHECK(!is_cpu_, "CPU non-scalar tensor is not supported!");
     }
 
+    if (!stride_order_.empty()) {
+      std::vector<IterDomain*> allocation_domain(rank);
+      for (auto i : c10::irange(rank)) {
+        allocation_domain[rank - 1 - static_cast<int>(stride_order_[i])] =
```
Comment (rdspring1) [inline csrc/python_frontend/fusion_record.h:1284]: Is there any invariant we need to enforce on `stride_order`?
`NVF_ERROR(stride_order > 0 && stride_order < rank, "stride order must be between 0 and tensor rank.")`

**Review by rdspring1 — COMMENTED** [review]
LGTM. Probably should add some error tests if there isn't a functional use case right now.

Is it useful to specify the stride order with negative numbers? e.g., backward indexing

File: csrc/python_frontend/fusion_record.h (hunk 3 above)
Comment (jjsjann123) [reply csrc/python_frontend/fusion_record.h:1284]: we can support negative indexing. I'll add that.

**Comment by jjsjann123** [comment]
> Probably should add some error tests if there isn't a functional use case right now.

There are actually a few places where this code path is hit unintentionally. (with expanded/unsqueezed dimensions).

File: python_tests/pytest_input_generators.py [hunk 4]
```diff
@@ -955,28 +955,28 @@ def permute_error_generator(
...
     yield SampleInput(
         make_arg(input_shape), [0, 1, 1, 3]
-    ), RuntimeError, "Duplicate entries in transformation map"
+    ), RuntimeError, "duplicated dimension entries"
 
     # TODO Add in-range axis check.
     yield SampleInput(
         make_arg(input_shape), [0, 1, 2, 4]
-    ), RuntimeError, "New2Old axes are not within the number of dimensions of the provided domain"
+    ), RuntimeError, "dims argument is out of range, expects"
```
Comment (rdspring1) [inline python_tests/pytest_input_generators.py:963]: You can use `(.*?)` to match any string, which is useful for error handling checks.

**Comment by jjsjann123** [comment]
added pytest checks. time to take another look @rdspring1

**Comment by jjsjann123** [comment]
!build

**Review by rdspring1 — APPROVED** [review]
LGTM

**Comment by jjsjann123** [comment]
Note to myself. I need to bump the version number before merging anything.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
Note. I'm still holding back from merging this PR. There's quite a few failing tests that I patched in separate PRs. I'll have those merged before shipping this one.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
I am aware of an existing codegen failure from this PR. I'll merge it as-is since that failure isn't present on our CI yet.

I will open an issue with a python repro so we can continue investigation there.

**Comment by jjsjann123** [comment]
strange... I don't have the repro on my local machine.

**Comment by jjsjann123** [comment]
:cry: turns out the added python test is broken now.

Something changed since commit [9bb2da5](https://github.com/NVIDIA/Fuser/pull/1031/commits/9bb2da5b24c4e08a440c54069fb620bdf5d2a42a)

**Comment by jjsjann123** [comment]
ha, I believe the error is real... It's passing earlier probably because we aren't setting the contiguity/allocationDomain properly for the test before. I'm taking another look at this.

**Comment by jjsjann123** [comment]
I was blind. the test copied over was busted... patching that. Looking forward to have this PR merged after CI cleanr

**Comment by jjsjann123** [comment]
!build

**Pull request merged**